src/python/grpcio/grpc/_observability.py

```diff
@@ -0,0 +1,125 @@
+
+from grpc._cython import cygrpc as _cygrpc
+
+_LOGGER = logging.getLogger(__name__)
+
+_channel = Any  # _channel.py imports this module.
+ClientCallTracerCapsule = TypeVar('ClientCallTracerCapsule')
+ServerCallTracerFactoryCapsule = TypeVar('ServerCallTracerFactoryCapsule')
+
+_plugin_lock: threading.RLock = threading.RLock()
+_OBSERVABILITY_PLUGIN: Optional[_ObservabilityPlugin] = None  # pylint: disable=used-before-assignment
+
+
+class _ObservabilityPlugin(Generic[ClientCallTracerCapsule,
+                                   ServerCallTracerFactoryCapsule],
+                           metaclass=abc.ABCMeta):
+    """
+    Note: Any future methods added to this interface cannot have the @abc.abstractmethod annotation.
+    """
+    _tracing_enabled: bool = False
+    _stats_enabled: bool = False
+
+    @abc.abstractmethod
+    def create_client_call_tracer(
+            self, method_name: bytes) -> ClientCallTracerCapsule:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def delete_client_call_tracer(
+            self, client_call_tracer: ClientCallTracerCapsule) -> None:
+        # delte client call tracer have to be called on o11y package side.
+        # Call it for both segregated and integrated call (`_process_integrated_call_tag`)
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def save_trace_context(self, trace_id: str, span_id: str,
+                           is_sampled: bool) -> None:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def create_server_call_tracer_factory(
+            self) -> ServerCallTracerFactoryCapsule:
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def record_rpc_latency(self, method: str, rpc_latency: float,
+                           status_code: Any) -> None:
+        raise NotImplementedError()
+
+    def set_tracing(self, enable: bool) -> None:
+        self._tracing_enabled = enable
+
+    def set_stats(self, enable: bool) -> None:
+        self._stats_enabled = enable
+
+    @property
+    def tracing_enabled(self) -> bool:
+        return self._tracing_enabled
+
+    @property
+    def stats_enabled(self) -> bool:
+        return self._stats_enabled
+
+    @property
+    def observability_enabled(self) -> bool:
+        return self.tracing_enabled or self.stats_enabled
+
+
+@contextlib.contextmanager
+def get_plugin() -> Optional[_ObservabilityPlugin]:
+    with _plugin_lock:
+        yield _OBSERVABILITY_PLUGIN
+
+
+def set_plugin(observability_plugin: Optional[_ObservabilityPlugin]) -> None:
+    global _OBSERVABILITY_PLUGIN  # pylint: disable=global-statement
+    with _plugin_lock:
+        if observability_plugin and _OBSERVABILITY_PLUGIN:
+            raise ValueError("observability_plugin was already set!")
+        _OBSERVABILITY_PLUGIN = observability_plugin
+
+
+def _observability_init(observability_plugin: _ObservabilityPlugin) -> None:
+    set_plugin(observability_plugin)
```

Comment:
We don't want to allow nested context managers, so we should raise an exception if we try to set a plugin when one has already been installed.